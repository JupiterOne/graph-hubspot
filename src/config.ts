import {
  IntegrationExecutionContext,
  IntegrationInstanceConfig,
  IntegrationInstanceConfigFieldMap,
  IntegrationValidationError,
} from '@jupiterone/integration-sdk-core';
import { createAPIClient } from './client';

export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  oauthAccessToken: {
    type: 'string',
    mask: true,
  },
  apiBaseUrl: {
    type: 'string',
    mask: true,
  },
};

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  /**
   * This access_token is considered valid as a Bearer Authorization header
   */
  oauthAccessToken: string;

  /**
   * Hubspot API base url
   */
  apiBaseUrl: string;
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationConfig>,
) {
  const { config } = context.instance;
  if (!Object.keys(instanceConfigFields).every((key) => config[key])) {
    throw new IntegrationValidationError(
      `Config requires all of {${Object.keys(instanceConfigFields).join(
        ', ',
      )}}`,
    );
  }

  const apiClient = createAPIClient(config, context.executionHistory);
  await apiClient.verifyAuthentication();
}
