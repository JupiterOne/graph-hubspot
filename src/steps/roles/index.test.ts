import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';
import { createIntegrationConfig } from '../../../test/config';
import { IntegrationConfig } from '../../config';
import * as role from '.';

describe('#fetchRoles', () => {
  // TODO: we'll add recordings later
  // let recording: Recording;

  // beforeEach(() => {
  //   recording = setupHubSpotRecording({
  //     directory: __dirname,
  //     name: 'fetchRoles',
  //   });
  // });

  // afterEach(async () => {
  //   await recording.stop();
  // });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: createIntegrationConfig(),
    });

    await role.fetchRoles(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    const roles = context.jobState.collectedEntities.filter((e) =>
      e._class.includes('AccessRole'),
    );

    expect(roles.length).toBeGreaterThan(0);
    expect(roles).toMatchGraphObjectSchema({
      _class: ['AccessRole'],
      schema: {
        additionalProperties: false,
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: 'hubspot_role' },
          // We want to make sure we include every field from converter.ts
          name: { type: 'string' },
          requiresBillingWrite: { type: 'boolean' },
        },
      },
    });
  });
});
