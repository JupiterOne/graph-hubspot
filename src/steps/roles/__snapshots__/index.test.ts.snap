// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchRoles should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "AccessRole",
      ],
      "_key": "hubspot_role:97200 ",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "97200",
            "name": "ExampleRole",
            "requiresBillingWrite": false,
          },
        },
      ],
      "_type": "hubspot_role",
      "createdOn": undefined,
      "displayName": "ExampleRole",
      "id": "97200",
      "name": "ExampleRole",
      "requiresBillingWrite": false,
    },
    Object {
      "_class": Array [
        "AccessRole",
      ],
      "_key": "hubspot_role:97591 ",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "97591",
            "name": "ExampleRole2",
            "requiresBillingWrite": false,
          },
        },
      ],
      "_type": "hubspot_role",
      "createdOn": undefined,
      "displayName": "ExampleRole2",
      "id": "97591",
      "name": "ExampleRole2",
      "requiresBillingWrite": false,
    },
  ],
  "collectedRelationships": Array [],
  "encounteredTypes": Array [
    "hubspot_role",
  ],
  "numCollectedEntities": 2,
  "numCollectedRelationships": 0,
}
`;
